### PR TITLE
Add support for RSA_SIGN_PKCS1_4096_SHA256 - fixes #16

### LINKED ISSUE #16: RSA_SIGN_PKCS1_4096_SHA256 supported?
Hello,

I am trying to sign a binary with our certificate that was created with "4096 bit RSA - PKCS#1 v1.5 padding - SHA256 Digest", but SignTool gives me `SignTool Error: No private key is available.`

is this supported?  

In the docs under limitations we have:
 * The purpose for the CryptoKey is ASYMMETRIC_SIGN.
 * The protection level for the CryptoKeyVersion is HSM.

Should that also state that RSA keys are not supported? I overlooked the section above that talks about ECDSA signing, but it's not clear if the single supported algo listed there is a limitation of EC Signing, or Signing altogether?

I guess i will have to get my certificate reissued for a private key with ECDSA on the P-256 Curve with a SHA-256 digest?

It shouldn't be a problem that the intermediate certificates are using RSA?

```
C:\certificate>certutil -repairstore My xxx
My "Personal"
================ Certificate 9 ================
Serial Number: xxx
Issuer: CN=DigiCert Trusted G4 Code Signing RSA4096 SHA384 2021 CA1, O=DigiCert, Inc., C=US
 NotBefore: 04/08/2023 00:00
 NotAfter: 31/05/2025 23:59
Subject: CN=Commercebuild Holdings Inc., O=Commercebuild Holdings Inc., L=Burnaby, S=British Columbia, C=CA
Non-root Certificate
Cert Hash(sha1): xxx
I0804 10:15:58.264925    5864 object_loader.cc:52] INFO: version projects/xm-cloudbuild/locations/us/keyRings/build-ring/cryptoKeys/bc-connector-sign-1/cryptoKeyVersions/1 is not loadable due to unsupported algorithm RSA_SIGN_PKCS1_4096_SHA256
```

### PR DESCRIPTION
**Pull request by davoustp**
Bringing support for RSA_SIGN_PKCS1_4096_SHA256 crypto algorithm.

Successfully signed a Windows artifact using the locally built provider, digital signature validates OK.

Note: documentation is currently not updated with this PR.

Please feel free to ask / amend / request changes. :-)

### CODE REVIEW

File: kmscng/operation/sign_utils.cc [hunk 1]
```diff
@@ -80,6 +185,8 @@ absl::StatusOr<const EVP_MD*> DigestForAlgorithm(
       return EVP_sha256();
     case kms_v1::CryptoKeyVersion::EC_SIGN_P384_SHA384:
       return EVP_sha384();
+    case kms_v1::CryptoKeyVersion::RSA_SIGN_PKCS1_4096_SHA256:
```
[the remark this answers is not in the document]
Comment (davoustp) [reply kmscng/operation/sign_utils.cc]: Done

File: kmscng/operation/sign_utils.cc [hunk 2]
```diff
@@ -57,13 +56,119 @@ absl::Status CopySignature(Object* object, std::string_view src,
...
+        SOURCE_LOCATION);
+  }
+  const uint8_t* sig_data = reinterpret_cast<const uint8_t*>(src.data());
+  std::copy(sig_data, sig_data + src.size(), dest.data());
+  return absl::OkStatus();
+}
+
+absl::Status CopySignature(Object* object, std::string_view src,
+                           absl::Span<uint8_t> dest) {
+  RETURN_IF_ERROR(IsValidSigningAlgorithm(object->algorithm()));
```
[the remark this answers is not in the document]
Comment (davoustp) [reply kmscng/operation/sign_utils.cc]: Agreed, done.

File: kmscng/operation/sign_utils.cc [hunk 3]
```diff
@@ -57,13 +56,119 @@ absl::Status CopySignature(Object* object, std::string_view src,
...
+  std::vector<uint8_t> result(header_size + payload_size);
+  // Header
+  BCRYPT_RSAKEY_BLOB* header =
+      reinterpret_cast<BCRYPT_RSAKEY_BLOB*>(result.data());
+  ASSIGN_OR_RETURN(header->Magic, MagicIdForAlgorithm(object->algorithm()));
+  header->BitLength = RSA_bits(rsa);
+  header->cbPublicExp = public_exponent_size;
+  header->cbModulus = modulus_size;
+  header->cbPrime1 = 0L; // not used by public keys
+  header->cbPrime2 = 0l; // not used by public keys
```
[the remark this answers is not in the document]
Comment (davoustp) [reply kmscng/operation/sign_utils.cc]: :-)

File: kmscng/algorithm_details.cc [hunk 4]
```diff
@@ -59,6 +60,13 @@ static const auto* const kAlgorithmDetails =
             BCRYPT_ECDSA_P384_ALGORITHM,  // algorithm_property
             NCRYPT_ALLOW_SIGNING_FLAG,    // key_usage
         },
+        {
+            kms_v1::CryptoKeyVersion::RSA_SIGN_PKCS1_4096_SHA256,  // algorithm
+            kms_v1::CryptoKey::ASYMMETRIC_SIGN,             // purpose
+            NCRYPT_RSA_ALGORITHM_GROUP,                     // algorithm_group
+            BCRYPT_RSA_ALGORITHM,         // algorithm_property
+            NCRYPT_ALLOW_SIGNING_FLAG,    // key_usage
+        },
```
[the remark this answers is not in the document]
Comment (davoustp) [reply kmscng/algorithm_details.cc:69]: Missed the provider.h enum - now added.

**Comment by davoustp** [comment]
> Thanks again for this PR! All minor comments, do you mind also updating kmscng/main/bridge_test.cc? Mostly just interested in RSA success tests for OpenKey, ExportKey, and SignHash.

Sure, no problem, done.

> Other than that it's pretty much ready to be merged, I'll play around some more with this as soon as I have some time to do the final manual testing pass.

Let me know how that goes.

**Comment by davoustp** [comment]
Just committed the doc update in kmscng/docs to add RSA support details.
Let me know if you want me to squash all those commits together.

**Comment by tdbhacks** [comment]
Alright, tested manually and everything looks good to me!

> Just committed the doc update in kmscng/docs to add RSA support details. Let me know if you want me to squash all those commits together.

Thanks for adding docs too, and yes please, happy to merge this PR once you have squashed and rebased (I've pushed a couple internal commits that had not made their way to HEAD yet, they shouldn't cause merge conflicts with your changes though). Thanks! :)

**Comment by davoustp** [comment]
> Alright, tested manually and everything looks good to me!
Nice!

> Thanks for adding docs too, and yes please, happy to merge this PR once you have squashed and rebased (I've pushed a couple internal commits that had not made their way to HEAD yet, they shouldn't cause merge conflicts with your changes though). Thanks! :)

Done.
BTW, do you already have an ETA for next release?

Thanks for the quick turnaround, btw, it's always a pleasure to contribute in this situation ;-)

**Review by tdbhacks — APPROVED** [review]

**Comment by tdbhacks** [comment]
> Done. BTW, do you already have an ETA for next release?

Left a comment on the open issue for greater visibility.

> Thanks for the quick turnaround, btw, it's always a pleasure to contribute in this situation ;-)

No problem, thanks again for being the first contributor of this repo! :)

**Pull request merged**